protocol/0082-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,238 @@
+- all `referral_reward_factor` values must be greater than or equal to `0` and less than or equal to the network parameter `referralProgram.maxReferralRewardFactor`.
+- all `referral_discount_factor` values must be greater than or equal to `0` and be less than or equal to the network parameter `referralProgram.maxReferralDiscountFactor`.
+- `window_length` must be an integer strictly greater than zero.
+
+The referral program will start the epoch after the `enactment_timestamp` is reached.
+
+## Referral Program Lifecycle:
+
+After a referral program [proposal](#governance-proposals) is validated and accepted by the network, the network referral program is created / updated and can be one of the following states. The current state of the network referral program should be exposed via an API.
+
+| Status               | Benefits Enabled | Condition for entry                                       | Condition for exit                                                |
+| -------------------- | ---------------- | --------------------------------------------------------- | ----------------------------------------------------------------- |
+| `STATUS_PROPOSED`    | No               | Governance proposal valid and accepted                    | Governance proposal voting period ends (or proposal is invalid)   |
+| `STATUS_REJECTED`    | No               | Governance vote fails (or is invalid)                     | New governance proposal submitted to the network                 |
+| `STATUS_PENDING`     | No               | Governance vote passes                                    | End of epoch after network reaches proposal `enactment_timestamp` |
+| `STATUS_ACTIVE`      | Yes              | Previously `STATUS_PENDING`                               | End of epoch after network reaches proposal `closing_timestamp`   |
+| `STATUS_CLOSED`      | No               | Previously `STATUS_ACTIVE`                                | New governance proposal submitted to the network                  |              
+
+## Team Mechanics
+A team is comprised of a referrer and all their referees. There can only ever be one referrer per team but the number of referees is unlimited.
+
+### Creating / updating a team
+
+To create a team and generate a referral code, a party must fulfil the following criteria:
+
+- party must not currently be a **referrer**
+- party must not currently be a **referee**
+- party must be staking at least `referralProgram.minStakedVegaTokens` tokens.
+- party must not have a liquidity provision in any of the following states:
+    - `STATUS_ACTIVE`
+    - `STATUS_PENDING`
+    - `STATUS_UNDEPLOYED`
+
+This staking requirement is constant. If a referrer un-stakes enough tokens to fall below the requirement, they and their referees will no long be eligible for program benefits. If the referrer re-stakes enough tokens to fulfil the staking requirement, the team will become eligible for referral program benefits.
```

Comment:
Do we want a referrer un-staking to punish referees also?

Or should we do something more along the lines of, set `referral_reward_factor` of all referees to `0` until the staking requirement is met (i.e. they receive no rewards).